From adfe144533960dc3513eadf8bf754afd37e07425 Mon Sep 17 00:00:00 2001
From: Nanxin Qin <nanxin.qin@amlogic.com>
Date: Thu, 7 Jan 2021 16:51:02 +0800
Subject: [PATCH 2/2] v4l: fixed stuck after seek if the first frame is not
 keyframe. [1/1]

PD#SWPL-39190

Problem:
V4L2 decoder can start with non key frame

Solution:
fixed decoding stuck after seek if frist frame isn't keyframe,
there is detail is dropping non I frame until reciver keyframe.

Verify:
ah212

Change-Id: I7465bdba0e9c32a55470d7540dd952959a4f3452
Signed-off-by: Nanxin Qin <nanxin.qin@amlogic.com>
---
 drivers/frame_provider/decoder/vav1/vav1.c | 11 +++++++++--
 drivers/frame_provider/decoder/vp9/vvp9.c  |  9 +++++++++
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/frame_provider/decoder/vav1/vav1.c b/drivers/frame_provider/decoder/vav1/vav1.c
index 3ff46e41..12680b96 100644
--- a/drivers/frame_provider/decoder/vav1/vav1.c
+++ b/drivers/frame_provider/decoder/vav1/vav1.c
@@ -7881,6 +7881,7 @@ static irqreturn_t vav1_isr_thread_fn(int irq, void *data)
 	struct AV1HW_s *hw = (struct AV1HW_s *)data;
 	unsigned int dec_status = hw->dec_status;
 	int obu_type;
+    int ret = 0;
 
 	/*if (hw->wait_buf)
 	 *	pr_info("set wait_buf to 0\r\n");
@@ -8259,12 +8260,18 @@ static irqreturn_t vav1_isr_thread_fn(int irq, void *data)
 		}
 	}
 
-	av1_continue_decoding(hw, obu_type);
+	ret = av1_continue_decoding(hw, obu_type);
 	hw->postproc_done = 0;
 	hw->process_busy = 0;
 
 	if (hw->m_ins_flag) {
-		start_process_time(hw);
+		if (ret >= 0)
+			start_process_time(hw);
+		else {
+			hw->dec_result = DEC_RESULT_DONE;
+			amhevc_stop();
+			vdec_schedule_work(&hw->work);
+		}
 	}
 
 	return IRQ_HANDLED;
diff --git a/drivers/frame_provider/decoder/vp9/vvp9.c b/drivers/frame_provider/decoder/vp9/vvp9.c
index 2f664acd..cc2b64da 100644
--- a/drivers/frame_provider/decoder/vp9/vvp9.c
+++ b/drivers/frame_provider/decoder/vp9/vvp9.c
@@ -8760,6 +8760,15 @@ static irqreturn_t vvp9_isr_thread_fn(int irq, void *data)
 		pbi->frame_width = pbi->vp9_param.p.width;
 		pbi->frame_height = pbi->vp9_param.p.height;
 
+		if (!pbi->has_keyframe &&
+			((pbi->frame_width == 0) ||
+			(pbi->frame_height == 0))) {
+			continue_decoding(pbi);
+			pbi->postproc_done = 0;
+			pbi->process_busy = 0;
+			return IRQ_HANDLED;
+		}
+
 		if (!v4l_res_change(pbi)) {
 			if (ctx->param_sets_from_ucode && !pbi->v4l_params_parsed) {
 				struct aml_vdec_ps_infos ps;
-- 
2.31.0

